server/db.go

```diff
@@ -419,49 +642,198 @@ func (c *Server) rotateIssuers() (err error) {
+		tx.Rollback()
+		return fmt.Errorf("failed to close rows on v3 issuer creation: %w", err)
 	}
+	queryTimer.ObserveDuration()
+	return tx.Commit()
+}
 
-	queryTimer := prometheus.NewTimer(createIssuerDBDuration)
-	rows, err := c.db.Query(
-		`INSERT INTO issuers(issuer_type, issuer_cohort, signing_key, max_tokens, expires_at, version) VALUES ($1, $2, $3, $4, $5, 2)`,
-		issuerType,
-		issuerCohort,
-		signingKeyTxt,
-		maxTokens,
-		expiresAt,
+// on the transaction, populate v3 issuer keys for the v3 issuer
+func txPopulateIssuerKeys(logger *logrus.Logger, tx *sqlx.Tx, issuer Issuer) error {
+	var (
+		duration *timeutils.ISODuration
+		err      error
 	)
+
+	if issuer.Version == 3 {
+		// get the duration from the issuer
+		duration, err = timeutils.ParseDuration(issuer.Duration)
+		if err != nil {
+			return fmt.Errorf("failed to parse issuer duration: %w", err)
+		}
+	}
+
+	// v1/v2 issuers only have a buffer of 1
+	if issuer.Version < 3 {
+		issuer.Buffer = 1
+	}
+
+	start := issuer.ValidFrom
+	i := 0
+	// time to create the keys associated with the issuer
+	if issuer.Keys == nil || len(issuer.Keys) == 0 {
+		issuer.Keys = []IssuerKeys{}
+	} else {
+		// if the issuer has keys already, start needs to be the last item in slice
+		start = issuer.Keys[len(issuer.Keys)-1].EndAt
+		i = len(issuer.Keys)
+	}
+
+	valueFmtStr := ""
+
+	var keys = []issuerKeys{}
+
+	// for i in buffer, create signing keys for each
+	for ; i < issuer.Buffer; i++ {
+		end := new(time.Time)
+		if duration != nil {
+			// start/end, increment every iteration
+			end, err = duration.From(*start)
+			if err != nil {
+				tx.Rollback()
+				return fmt.Errorf("unable to calculate end time: %w", err)
+			}
+		}
+
+		signingKey, err := crypto.RandomSigningKey()
+		if err != nil {
+			logger.Error("Error generating key")
+			tx.Rollback()
+			return err
+		}
+
+		signingKeyTxt, err := signingKey.MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling signing key")
+			tx.Rollback()
+			return err
+		}
+
+		pubKeyTxt, err := signingKey.PublicKey().MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling public key")
+			tx.Rollback()
+			return err
+		}
+
+		keys = append(keys, issuerKeys{
+			SigningKey: signingKeyTxt,
+			PublicKey:  string(pubKeyTxt),
+			Cohort:     issuer.IssuerCohort,
+			IssuerID:   issuer.ID,
+			StartAt:    start,
+			EndAt:      end,
+		})
+
+		if start != nil && !(*start).Equal(*issuer.ValidFrom) {
+			valueFmtStr += ", "
+		}
+		valueFmtStr += "($1, $2, $3, $4, $5, $6)"
+
+		// increment start
+		if start != nil && end != nil {
+			*start = *end
+		}
+	}
+
+	var values []interface{}
+	// create our value params for insertion
+	for _, v := range keys {
+		values = append(values,
+			v.IssuerID, v.SigningKey, v.PublicKey, v.Cohort, v.StartAt, v.EndAt)
+	}
+
+	rows, err := tx.Query(
+		fmt.Sprintf(`
+		INSERT INTO v3_issuer_keys
+			(
+				issuer_id,
+				signing_key,
+				public_key,
+				cohort,
```

Comment:
I think this could lead to an out of range as the v3_issuer_keys.cohort is a type small int and this value in an int. See comment on the migrations.

Reply:
will convert issuer cohort in code to int16 to match.